contracts/src/staking/README.md

```diff
@@ -0,0 +1,136 @@
+1. Inform the Staking Manager contract of the change. The Staking Manager contract will construct a Warp message attesting to the change in the validator set.
+2. Deliver the Warp message containing the validator set change to the P-Chain. The P-Chain will construct a Warp message acknowledging the change to the validator set.
+3. Deliver the Warp message containing the P-Chain acknowledgement back to the Staking Manager contract on the Subnet. The Staking Manager contract finalizes the validator set change.
+
+### Initializing a Validator Set
+
+If the Subnet has no validators registered on the P-Chain, then it will not be able to sign a Warp message to add a validator (step 1 above). In this case, Subnet recovery must be performed on the P-Chain, which allows an initial validator set to be specified. The P-Chain constructs a Warp message (format specification TBD) describing the initial validator set (step 2 above).
+
+Once the Subnet has been recovered, the stake specified on the P-Chain must be locked in the Staking Contract before any further validator set changes may be made. This is done via the `provideInitialStake` method, and guarded by the `onlyWhenInitialStakeProvided` modifier.
+
+
+### Registering a Validator
+
+Validator registration is initiated with a call to `initializeValidatorRegistration` on the concrete Staking Manager contract. The Staking Manager contract constructs a [`RegisterSubnetValidatorMessage`](#registersubnetvalidatormessage) Warp message to be sent to the P-Chain. Each validator registration request includes all of the information needed to identify the validator and its stake weight, as well as an `expiry` timestamp before which the `RegisterSubnetValidatorMessage` must be delivered to the P-Chain.
+
+The `RegisterSubnetValidatorMessage` is delivered to the P-Chain as the payload of a `RegisterSubnetValidatorTx`. Please see the transaction [specification](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#step-2-issue-a-registersubnetvalidatortx-on-the-p-chain) for validity requirements. The successful `RegisterSubnetValidatorTx` results in a [`SubnetValidatorRegistrationMessage`](#subnetvalidatorregistrationmessage) Warp message indicating that the specified validation was successfully registered on the P-Chain.
+
+The `SubnetValidatorRegistrationMessage` is delivered to the Staking Manager contract via a call to `completeValidatorRegistration`. Staking rewards begin accruing at this time.
+
+#### Handling a Missed Expiry
+
+In the case of a missed expiry, the `RegisterSubnetValidatorTx` will result in a `SubnetValidatorRegistrationMessage` Warp message with the `valid` field set to `0`. This serves as proof that the corresponding validation has not started and may never start. The `SubnetValidatorRegistrationMessage` can be provided to the Staking Manager contract by calling `completeEndValidation` with `setWeightMessageType=false`.
```

Comment:
```suggestion
In the case of a missed expiry, the `RegisterSubnetValidatorTx` will result in a `SubnetValidatorRegistrationMessage` Warp message with the `valid` field set to `0`. This serves as proof that the corresponding validation has not started and will never start. The `SubnetValidatorRegistrationMessage` can be provided to the Staking Manager contract by calling `completeEndValidation` with `setWeightMessageType=false`.
```

Reply:
Do we know if delivering the `RegisterSubnetValidatorTx` to the p-chain is idempotent? We have "resend" functions for all our communications with the p-chain, does the p-chain have the same?

Reply:
Replay protection is provided by the `expiry` field in the `RegisterSubnetValidatorTx` Warp message. I need to think more about the conditions under which resending is necessary. Since the staking contracts represent the only path by which validators can be added, the semantics around validator churn are slightly different than the general case.

Reply:
The semantics of the `valid` bit guarantee that the corresponding validation _has not_ and _will_ never start if `valid=0`.

Reply:
I guess my question is - What happens on the p-chain if we deliver the `RegisterSubnetValidatorTx` twice? Does it return a valid `SubnetValidatorRegistrationMessage` in both cases?

Reply:
I'm not sure what will happen on the P-Chain in that case. However, a `SubnetValidatorRegistrationMessage` with `valid=0` should only mean that the validation in question is not active and can never become active, and should not reflect the `RegisterSubnetValidatorTx` transaction status otherwise.